File: trunk/Geometry/eta.cpp
#ifndef ETA_H_
#define ETA_H_

#include <map>
#include <new>
using namespace std;

#include "geoquant.h"
#include "geoquants.h"
#include "triposition.h"
#include "triangulation/triangulation.h"

EtaIndex* Eta::Index = NULL;

Eta::Eta( Edge& e ){}

void Eta::recalculate(){}

void Eta::remove() {
     deleteDependents();
     Index->erase(pos);
}
Eta::~Eta(){}

Eta* Eta::At( Edge& e ){
  TriPosition T( 1, e.getSerialNumber() );
  if( Index == NULL ) Index = new EtaIndex();
  EtaIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    Eta* val = new Eta( e );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void Eta::CleanUp(){
  if( Index == NULL ) return;
  EtaIndex::iterator iter;
  EtaIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    delete iter->second;
  }
    
  delete Index;
  Index = NULL;
}

#endif /* ETA_H_ */

File: Java/src/geo/Eta.java
package geo;

import java.util.HashMap;

import triangulation.Edge;


public class Eta extends Geoquant {
  private static HashMap<TriPosition, Eta> Index = new HashMap<TriPosition, Eta>();
  
  private Eta(Edge e) {
    super(e); // ALWAYS have to call this first
    
  }
  
  public static Eta At(Edge e) {
    TriPosition T = new TriPosition( e.getSerialNumber() );
    Eta q = Index.get(T);
    if(q == null) {
      q = new Eta(e);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Edge e) {
    return At(e).getValue();
  }
  
  protected void recalculate() {
    // This is empty for an eta
  }

 
  public void remove() {
    deleteDependents();
    Index.remove(pos);
  }

}